### PR TITLE
Dedent fixes for Python 3.12

### PR DESCRIPTION
**Pull request by Harsha-Nori**
Aiming to fix https://github.com/guidance-ai/guidance/issues/942 

This should handle general use cases across Python versions, but further fixes are probably still needed for:

- closure functions
- nested f-strings (with arbitary nesting now possible in python 3.12 thanks to the new python parser improvements 😵‍💫)

I think these are both esoteric enough that we can kick the can down the road, but I'll keep thinking about it. 

@mmoskal

### CODE REVIEW

**Comment by Harsha-Nori** [comment]
Also tag @riedgar-ms -- I may have put the tests in the wrong place, I saw "test_call_embeddings" inside of "test_model" which feels spiritually similar to test semantics, but I'm really not sure where the right place is in the new framework. Happy to move these anywhere really

**Comment by riedgar-ms** [comment]
Looks like a reasonable location to me.

**Comment by hudson-ai** [comment]
Nice @Harsha-Nori! The solution is also a lot more readable than what was there before.

Regarding location, I think it's fine where it is, but I'd consider making a new file `test_decorator.py` or something of that sort in `tests/unit/`. If we later want to add some tests about inferring stateful/statelessness, etc. I think that we'd want to put those in there too.

File: guidance/_utils.py [hunk 1]
```diff
@@ -91,39 +72,30 @@ def __exit__(self, exc_type, exc_value, traceback):
...
-        lines[1] = (
-            "    "
-            + ",".join(f.__code__.co_freevars)
-            + " = "
-            + ",".join("None" for _ in f.__code__.co_freevars)
-        )
-        source = "    \n".join(
-            lines
-        )  # TODO: this does not quite work because new_code_obj is now the __outer__closure_wrap() function...could be fixed with work...
+#       lines = source.split("\n")
```
Comment (riedgar-ms) [inline guidance/_utils.py:84]: Can these commented blocks (in the tests as well) be deleted before merge, please?

Comment (Harsha-Nori) [reply guidance/_utils.py:84]: Is it OK if we leave them in? The logic here is annoyingly tricky and while it's not feature complete yet, will be important to replicate when we extend the decorator to nested function definitions

**Comment by hudson-ai** [comment]
@riedgar-ms @Harsha-Nori any objections to merging this?

**Review by hudson-ai — APPROVED** [review]

**Review by riedgar-ms — APPROVED** [review]

**Comment by Harsha-Nori** [comment]
Thanks for approving @hudson-ai and @riedgar-ms!

**Pull request merged**